Write queue after connect windows

txdv opened the pull request.

txdv committed:
win: tcp,pipes - queue writes when yet not connected

txdv commented:
tty works, doesn't need a queue, because stdin/stdout is always open.

pipe on the other hand is a bit more difficult.

saghul commented on src/win/stream.c:
```diff
@@ -126,10 +126,6 @@ int uv_write(uv_write_t* req,
   uv_loop_t* loop = handle->loop;
   int err;
 
-  if (!(handle->flags & UV_HANDLE_WRITABLE)) {
```
we need to check if the handle is closing here and give an error.

saghul replied:
we also need to check if the handle is bound, aka, if handle->handle != INVALID_HANDLE. Unix returns EBADF in that case.

txdv replied:
It is tricky here.

tty doesn't need a queue because stdin/stdout is already open and creating a tty happens instantly (no callback/only init needed).

So we need to return broken pipe on stdin TTY.

Then we need to check if a handle is invalid, return EBADF, then we need to go on with the checks.

saghul commented on include/uv-win.h:
```diff
@@ -370,6 +370,9 @@ RB_HEAD(uv_timer_tree_s, uv_timer_s);
 
 #define UV_WRITE_PRIVATE_FIELDS                                               \
+  void* queue[2];                                                             \
+  uv_buf_t* bufs;                                                             \
```
Create a `bufsml` thing here like we do on Unix, to avoid malloc-ing if we have <= 4 buffers.

txdv replied:
This has been adressed

saghul commented on src/win/stream.c:
```diff
@@ -126,9 +126,11 @@ int uv_write(uv_write_t* req,
   int err;
 
-  if (!(handle->flags & UV_HANDLE_WRITABLE)) {
+  if (handle->type == UV_INVALID_HANDLE)
```
Should be:

```
if (handle->socket == INVALID_HANDLE_VALUE)
```

txdv replied:
crazy

txdv replied:
but stream handle doesn't have socket

saghul replied:
On 09/23/2014 11:43 AM, Andrius Bentkus wrote:

> In src/win/stream.c:
> 
> > @@ -126,9 +126,11 @@ int uv_write(uv_write_t\* req,
> >    uv_loop_t\* loop = handle->loop;
> >    int err;
> > -  if (!(handle->flags & UV_HANDLE_WRITABLE)) {
> > -  if (handle->type == UV_INVALID_HANDLE)
> 
> but stream handle doesn't have socket

That applies to TCP, pipes and ttys use handle->handle.

## 

Saúl Ibarra Corretgé
bettercallsaghul.com

txdv replied:
there is no such thing as UV_INVALID_HANDLE, the last patch though uses what I think you wanted

txdv commented:
Boom, rebase fellas.

@saghul please check the return error codes in uv_write

I want to merge tcp_write_after_connect and pipe_write_after_connect into write_after_connect and rename it to write_before_connect.

Furthermore i'm not happy with uv_pipe_connect not returning an int and returning all the error codes only in the callback. [This is some ancient code](https://github.com/joyent/libuv/blame/master/src/unix/pipe.c#L208-L214) and I think this should be different.

This also let me think a bit about the returning codes. We need to try to make it consistent that returning codes which are returned immediately or in the callback should be consistent on all platforms. uv_pipe_connect returned everything in the callbacks, which is easier to debug, but I don't feel like EINVAL on a bad handle input should be returned in the callback (it makes the code slower, because we have to run an entire loop tick just to return a bad handle error code).

saghul commented:
> Furthermore i'm not happy with uv_pipe_connect not returning an int and returning all the error codes only afterwards. This is some ancient code and I think this should be different.
> 
> This also let me think a bit about the returning codes. We need to try to make it consistent that returning codes which are returned immediately or in the callback should be consistent on all platforms. uv_pipe_connect returned everything in the callbacks, which is easier to debug, but I don't feel like EINVAL on a bad handle input should be returned in the callback (it makes the code slower, because we have to run an entire loop tick just to return a bad handle error code).

I don't disagree, but please don't derail into other issues here, just open a new one.

txdv committed:
add an assert afterwards

txdv committed:
group uv_buf_init with uv_loop_init (consistenty with the pipe test)

txdv committed:
style fixes

txdv committed:
test commit

txdv committed:
Make test beautiful.

txdv commented on src/win/stream.c:
```diff
@@ -126,9 +126,21 @@ int uv_write(uv_write_t* req,
+
+  /* brakes the test */
+  /*
+  if (handle->type == UV_NAMED_PIPE && ((uv_pipe_t*) handle)->handle == INVALID_HANDLE_VALUE)
```
This shit doesn't fly, because handle == INVALID_HANDLE_VALUE even though we use connect before issuing a write.

It gets != INVALID_HANDLE_VALUE only after the connect callback

txdv committed:
rename stuff for even nicer test

saghul commented on src/win/pipe.c:
```diff
@@ -1428,6 +1410,58 @@ static int uv_pipe_write_impl(uv_loop_t* loop,
+                              uv_write_cb cb) {
+  int size;
+
+  if (nbufs != 1 && (nbufs != 0 || !send_handle)) {
```
This test feels wrong, what if nbufs is 3?

txdv replied:
This is code from before, I don't know.

trevnorris replied:
Walking through the logic...

`uv_pipe_write` (which is called by `uv_write`) always sends a `NULL` `send_handle`, so all that matters is if `nbufs != 1`. Though the docs give an example of how to send `nbufs > 1` example. This seems to have two issues:
1. Incorrect documentation on the capabilities of using `uv_write` on pipes in Windows.
2. Why sending `nbufs == 0` returns `ERROR_NOT_SUPPORTED`? Seems that since there's technically no work to be done that it should just return immediately.

---

On the side, there's a logic flaw in `uv_write` in `src/win/stream.c`. Take the following snippet:

``` c
  err = ERROR_INVALID_PARAMETER;
  switch (handle->type) {
    /* case UV_{TCP,NAMED_PIPE,TTY}: */
    default:
      assert(0);
  }

  return uv_translate_sys_error(err);
```

Since `default: assert(0)` there's no sense in setting `err = ERROR_INVALID_PARAMETER;`

This same logic flaw exists in `uv_write2`

---

`uv_pipe_write2` is only called by `uv_write2` in `src/win/stream.c` and passes `send_handle`. Docs say:

> _send_handle_ must be a TCP socket or pipe, which is a server or a connection (listening or connected state). Bound sockets or pipes will be assumed to be servers.

In this case the check effectively becomes `if (nbufs > 1)`. The docs say about `uv_write2`:

> Extended write function for sending handles over a pipe. The pipe must be initialized with ipc == 1.

So I'm not sure why using `uv_write2` when `nbufs == 0` is fine, but not `uv_write`. And, again, there's the issue that only 1 `uv_buf_t` can be written at a time to a pipe on Windows.

saghul replied:
> Why sending nbufs == 0 returns ERROR_NOT_SUPPORTED? Seems that since there's technically no work to be done that it should just return immediately.

It should give ENOBUFS error.

saghul replied:
> only 1 uv_buf_t can be written at a time to a pipe on Windows.

We should be able to lift that, I'm not aware of any techinical limitations.

saghul replied:
> Since default: assert(0) there's no sense in setting err = ERROR_INVALID_PARAMETER;
> 
> This same logic flaw exists in uv_write2

In non-debug builds that assert is ignored by the compiler.

trevnorris replied:
@saghul 

> In non-debug builds that assert is ignored by the compiler.

I thought that only counts if `NDEBUG` is defined before `<assert.h>` is included, but I don't see that defined anywhere that would affect those two asserts.

saghul replied:
On 09/25/2014 07:07 PM, Trevor Norris wrote:

> In src/win/pipe.c:
> 
> > @@ -1428,6 +1410,58 @@ static int uv_pipe_write_impl(uv_loop_t\* loop,
> >    return 0;
> >  }
> > 
> > +static int uv_pipe_write_impl(uv_loop_t\* loop,
> > -                              uv_write_t\* req,
> > -                              uv_pipe_t\* handle,
> > -                              const uv_buf_t bufs[],
> > -                              unsigned int nbufs,
> > -                              uv_stream_t\* send_handle,
> > -                              uv_write_cb cb) {
> > -  int size;
> >   +
> > -  if (nbufs != 1 && (nbufs != 0 || !send_handle)) {
> 
> ```
> In non-debug builds that assert is ignored by the compiler.
> ```
> 
> I thought that only counts if |NDEBUG| is defined before |<assert.h>| is
> included, but I don't see that defined anywhere that would affect those
> two asserts.

It's usually defined by the compilation tool:

```
$ ./gyp_uv.py -f ninja
$ ninja -C out/Release -v
ninja: Entering directory `out/Release'
[1/166] cc -MMD -MF obj/test/run-tests.test-platform-output.o.d
-D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -D_GNU_SOURCE
-D_POSIX_C_SOURCE=200112 -DNDEBUG -I../../include -pthread -Wall -ansi
-pthread -fvisibility=hidden -O3 -fstrict-aliasing -fomit-frame-pointer
-fdata-sections -ffunction-sections   -c
../../test/test-platform-output.c -o
obj/test/run-tests.test-platform-output.o
...
```

saghul commented on src/win/pipe.c:
```diff
@@ -1742,6 +1776,8 @@ void uv_process_pipe_connect_req(uv_loop_t* loop, uv_pipe_t* handle,
     req->cb(req, uv_translate_sys_error(err));
   }
 
+  uv__stream_flush_write_queue((uv_stream_t*) handle, err ? UV_ECANCELED : 0, uv_pipe_write_impl_req);
```
I don't really like this approach. I'd rather have a function which sets the error value and puts all requests in the completed queue, and another one which runs the callbaclks. See `uv__stream_flush_write_queue` and `uv__write_callbacks` on Unix.

saghul commented on src/win/stream.c:
```diff
@@ -126,9 +126,21 @@ int uv_write(uv_write_t* req,
+
+  if (handle->flags & UV_HANDLE_LISTENING)
+    return UV_EINVAL;
+
```
I think  we need a better approach here. I'd say that we should return EPIPE if closing / closed and EINVAL if we are not trying to connect. "Trying to connect" would be when stream->connect_req is not NULL. We need to add that field on Windows, I think it makes things easier. Unix would also need adjusting. This is more strict, yes, but more correct and easier to reason about IMHO.

saghul commented on src/win/tcp.c:
```diff
@@ -1116,8 +1157,11 @@ void uv_process_tcp_connect_req(uv_loop_t* loop, uv_tcp_t* handle,
+
   req->cb(req, uv_translate_sys_error(err));
 
+  uv__stream_flush_write_queue((uv_stream_t*) handle, err ? UV_ECANCELED : 0, uv_tcp_write_req);
```
Same as above.

saghul commented on test/test-write-after-connect.c:
```diff
@@ -0,0 +1,146 @@
+  ASSERT(0 == uv_ip4_addr("127.0.0.1", TEST_PORT, &addr));
+
+  test(tcp_client, tcp_server);
+  test(pipe_client, pipe_server);
```
Better define 2 tests, one for TCP and one for pipes, I'd rather see them fail individually.

txdv replied:
2 TEST ENTRIES?

saghul replied:
Yes, I don't want write_after_connect if either the pipe or the tcp test failed, they should be separate. They can be on the same file, that's ok, but we need 2 TEST_IMPL things.

saghul commented:
Left a first round of comments.

saghul commented:
Thanks for working on this, Andrius.

txdv committed:
split uv__stream_flush_write_queue into two functions

looks probably nicer

txdv committed:
add two seperate tests.

txdv committed:
handle connection error values.

txdv commented:
> I think we need a better approach here. I'd say that we should return EPIPE if closing / closed and EINVAL if we are not trying to connect. "Trying to connect" would be when stream->connect_req is not NULL. We need to add that field on Windows, I think it makes things easier. Unix would also need adjusting. This is more strict, yes, but more correct and easier to reason about IMHO.

Another thing, how do we check for TCP?

shouldnt we better check with `handle->flags & UV_HANDLE_WRITABLE` instead that invalid handle value?

saghul commented:
> Another thing, how do we check for TCP?

Check what exactly?

> shouldnt we better check with handle->flags & UV_HANDLE_WRITABLE instead that invalid handle value?

Yes and no, that flag is only set after connet succeeds. So we need to determine if we are trying to connect beforehand.

txdv commented:
> Yes and no, that flag is only set after connet succeeds. So we need to determine if we are trying to connect beforehand.

Yeah, I already done that.

txdv committed:
connection/write check

saghul commented on include/uv-win.h:
```diff
@@ -438,7 +443,8 @@ RB_HEAD(uv_timer_tree_s, uv_timer_s);
     struct { uv_tcp_connection_fields };                                      \
-  };
+  };                                                                          \
+  int connect_req;
```
Use a pointer to the req, not an int.

txdv replied:
Yeah, you could specify that next time like totally immediately right in the first comment.

saghul replied:
I believe I did, though GitHub erased all comments. I explicitly mentioned Unix did it that way, and Unix doesn't use an int.

saghul commented on src/win/tcp.c:
```diff
@@ -1,4 +1,4 @@
-/* Copyright Joyent, Inc. and other Node contributors. All rights reserved.
+/* iopyright Joyent, Inc. and other Node contributors. All rights reserved.
```
?

txdv replied:
the lib does io man! fixed

saghul commented on src/win/tcp.c:
```diff
@@ -813,19 +816,22 @@ int uv_tcp_getpeername(const uv_tcp_t* handle,
-                 unsigned int nbufs,
-                 uv_write_cb cb) {
+int uv_tcp_write_req(uv_write_t* req)
+{
```
style: brace in the same line as function definition.

txdv replied:
style fixed.

saghul commented on src/win/stream-inl.h:
```diff
@@ -53,4 +53,39 @@ INLINE static void uv_connection_init(uv_stream_t* handle) {
 }
 
 
+INLINE static void uv__write_callbacks(uv_stream_t* stream, int (*callback)(uv_write_t* request)) {
```
I don't like passing the callback here. Please check the handle type and call the appropriate function instead. Also assert if we get the wrong handle type.

txdv replied:
The functions are defined as static functiosn in the files. I don't think I can access them?

saghul replied:
They don't need to be this way though, you can reorganize them a bit to make it better.

txdv replied:
as I said:

```
src/win/pipe.c:1187:12: error: static declaration of 'uv_pipe_write_impl_req' follows non-static declaration
 static int uv_pipe_write_impl_req(uv_write_t* req)
            ^
In file included from src/win/pipe.c:31:0:
src/win/stream-inl.h:70:9: note: previous implicit declaration of 'uv_pipe_write_impl_req' was here
         uv_pipe_write_impl_req(req);
```

txdv replied:
I feel like callback is the right solution here, because with your approach we are scattering the implementation details everywhere.

This way we have the logic right where it belongs, in pipe.c and tcp.c

how do you want me to reorganize them?

saghul replied:
I disagree. Look at what `uv_listen` does in stream.c. I find something like that better.

txdv replied:
Everytime we run a TCP connect callback, we will check if the queued write callbacks do not belong to a PIPE. ¯_(ツ)_/¯

Maybe we should really introduce UV_SAGHUL.

saghul replied:
No. Every time a connect callback fires we call the appropriate function to process it depending on the type of handle.

> Maybe we should really introduce UV_SAGHUL.

I'm starting to get tired of all you stupid and snarky remarks. If you don't like my approach you have 2 choices: prove me wrong, or get someone else to help you.

txdv replied:
sorry dude, don't take them too seriously, I don't want to insult your or anything, this is just a playful thing like writing those emoticons

https://github.com/txdv/libuv/commit/2674963879e819c380dc3f3ceaa1812fa9b721be

Is this the way you want it?

saghul replied:
> sorry dude, don't take them too seriously, I don't want to insult your or anything, this is just a playful thing like writing those emoticons

I guess we are just different. I'm not mad, but please let's just focus on the code.

txdv committed:
use connect_req pointer

txdv committed:
I thought iopyright was appropriate because the lib does io, but saghul disagrees as usual.

txdv committed:
style fix

saghul commented:
Please leave a nothe when you're done addressing comments, GH doesn't notify me when you add new commits.

txdv commented:
it removes the once addressed from this view, i'll add a comment anyway.

txdv committed:
¯\_(ツ)_/¯

saghul commented on src/win/stream.c:
```diff
@@ -126,9 +126,25 @@ int uv_write(uv_write_t* req,
   int err;
 
-  if (!(handle->flags & UV_HANDLE_WRITABLE)) {
+  if (handle->flags & UV_HANDLE_CLOSED || handle->flags & UV__HANDLE_CLOSING)
```
uv__is_closing

txdv replied:
```
In file included from src/win/internal.h:26:0,
                 from src/win/stream.c:25:
src/win/stream.c: In function 'uv_write':
src/win/../uv-common.h:149:19: error: 'UV_CLOSING' undeclared (first use in this function)
   (((h)->flags & (UV_CLOSING |  UV_CLOSED)) != 0)
                   ^
src/win/stream.c:129:43: note: in expansion of macro 'uv__is_closing'
   if (handle->flags & UV_HANDLE_CLOSED || uv__is_closing(handle))
                                           ^
src/win/../uv-common.h:149:19: note: each undeclared identifier is reported only once for each function it appears in
   (((h)->flags & (UV_CLOSING |  UV_CLOSED)) != 0)
                   ^
src/win/stream.c:129:43: note: in expansion of macro 'uv__is_closing'
   if (handle->flags & UV_HANDLE_CLOSED || uv__is_closing(handle))
                                           ^
src/win/../uv-common.h:149:33: error: 'UV_CLOSED' undeclared (first use in this function)
   (((h)->flags & (UV_CLOSING |  UV_CLOSED)) != 0)
                                 ^
src/win/stream.c:129:43: note: in expansion of macro 'uv__is_closing'
   if (handle->flags & UV_HANDLE_CLOSED || uv__is_closing(handle))
                                           ^
```

Even though it is common it is only being used in unix.

trevnorris replied:
Is this an issue that should be addressed? The flags are the same:

`src/unix/internal.h`:

``` c
/* handle flags */
enum {
  UV_CLOSING              = 0x01,   /* uv_close() called but not finished. */
  UV_CLOSED               = 0x02,   /* close(2) finished. */
```

`src/win/internal.h`

``` c
#define UV_HANDLE_CLOSED                        0x00000002
/* uv-common.h: #define UV__HANDLE_CLOSING      0x00000001 */
```

Seems the difference is one uses an `enum` and the other uses `#define`'s. But the flag values are the same. So you could make an alias of `UV_CLOSED` and `UV_CLOSING` in `src/win/internal.h` and use `uv__is_closing` for consistency.

saghul replied:
It should, though not in this patch. We have `uv_is_closing` on Windows.

saghul commented on src/win/stream.c:
```diff
@@ -126,9 +126,25 @@ int uv_write(uv_write_t* req,
+  if (handle->type == UV_TTY && ((uv_tty_t*) handle)->handle == INVALID_HANDLE_VALUE)
+    return UV_EINVAL;
+
+  if (handle->flags & UV_HANDLE_LISTENING)
```
Looks like we no longer need this, it would be covered by the above test.

saghul commented on src/win/pipe.c:
```diff
@@ -1180,40 +1184,20 @@ static void uv_queue_non_overlapped_write(uv_pipe_t* handle) {
-                              uv_stream_t* send_handle,
-                              uv_write_cb cb) {
+int uv_pipe_write_impl_req(uv_write_t* req)
+{
```
style

txdv replied:
man, these are so tedious to deal with, maybe we should use a C lint checker?

txdv replied:
fixed

txdv committed:
style fix.

saghul commented on src/win/tcp.c:
```diff
@@ -1116,8 +1159,14 @@ void uv_process_tcp_connect_req(uv_loop_t* loop, uv_tcp_t* handle,
   req->cb(req, uv_translate_sys_error(err));
 
+  if (!err)
+    uv__write_callbacks((uv_stream_t*) handle);
```
We need a better name for this, since it doesn't fire any callback. Maybe `uv__stream_write`.

txdv replied:
uv__execute_write_callbacks ? What name do you want?

saghul replied:
how about `uv__run_write_callbacks`? I think we use similar naming somewhere internally.

txdv replied:
Now we have:

``` c
uv__run_write_callbacks((uv_stream_t*) handle);
uv__stream_flush_write_queue((uv_stream_t*) handle, UV_ECANCELED);
```

Shouldn't they be more consistent?

saghul commented on src/win/stream-inl.h:
```diff
@@ -53,4 +53,43 @@ INLINE static void uv_connection_init(uv_stream_t* handle) {
+        uv_tcp_write_req(req);
+      else if (stream->type == UV_NAMED_PIPE)
+        uv_pipe_write_impl_req(req);
+
```
else assert.

txdv replied:
done

saghul commented on src/win/stream-inl.h:
```diff
@@ -53,4 +53,43 @@ INLINE static void uv_connection_init(uv_stream_t* handle) {
+
+    if (req != NULL) {
+      if (stream->type == UV_TCP)
+        uv_tcp_write_req(req);
```
Error handling needs to be re-checked. This function returns an error if it fails, which we are ignoring here.

txdv replied:
done

txdv committed:
error checking and assert

txdv committed:
No need for this error check.

trevnorris commented on src/win/pipe.c:
```diff
@@ -1180,40 +1184,19 @@ static void uv_queue_non_overlapped_write(uv_pipe_t* handle) {
-  memset(&req->overlapped, 0, sizeof(req->overlapped));
+  uv_loop_t* loop = req->handle->loop;
+  uv_pipe_t* handle = (uv_pipe_t*) req->handle;
+  const uv_buf_t *bufs = req->bufs;
```
style nit: `chat uv_buf_t* bufs`

trevnorris commented on src/win/pipe.c:
```diff
@@ -1428,6 +1411,58 @@ static int uv_pipe_write_impl(uv_loop_t* loop,
+  if (handle->handle == INVALID_HANDLE_VALUE) {
+    if (nbufs > ARRAY_SIZE(req->bufsml)) {
+      req->bufs = malloc(size);
+      if (req->bufs == 0)
```
Just curious, wouldn't `if (req->bufs == NULL)` be more appropriate?

txdv replied:
should we use `!req->bufs` or `req->bufs == NULL`?

saghul replied:
I personally prefer to explicitly check for NULL.

trevnorris commented on src/win/pipe.c:
```diff
@@ -1428,6 +1411,58 @@ static int uv_pipe_write_impl(uv_loop_t* loop,
+  } else {
+    req->bufs = (uv_buf_t*) bufs;
+    return uv_pipe_write_impl_req(req);
+  }
```
The indentation could be simplified if it were just:

``` c
if (handle->handle != INVALID_HANDLE_VALUE) {
  req->bufs = (uv_buf_t*) bufs;
  return uv_buf_write_impl_req(req);
}
```

Then the rest could be flattened.

Though I'm not understanding the use of `INVALID_HANDLE_VALUE` here. Since `uv_pipe_write_impl_req` must be called if the "pipe [was] initialized with ipc == 1".

txdv replied:
The handle->handle gets only assigned AFTER the iocp connect callback.

If we do a write before, we queue it internally.

This is to ensure that this works:

``` c
connect
write
write
```

because right now we have to do something liek this:

``` c
connect ->
  write
  write
```

where -> is the functionality in the callback

saghul replied:
I'd rather see the check like: `if (handle->connect_req)`, it's more obvious.

trevnorris commented on src/win/pipe.c:
```diff
@@ -1742,6 +1779,11 @@ void uv_process_pipe_connect_req(uv_loop_t* loop, uv_pipe_t* handle,
+  if (!err)
+    uv__write_callbacks((uv_stream_t*) handle);
+  else
+    uv__stream_flush_write_queue((uv_stream_t*) handle, UV_ECANCELED);
```
According to @piscisaureus the data should be written before `req->cb()` is called:

> <piscisaureus> what you should do is start flushing the write queue in uv_tcp_process_connect_req
> <piscisaureus> just insert the logic here -> https://github.com/joyent/libuv/blob/master/src/win/tcp.c#L1112
> <piscisaureus> [Y]ou have to consider the case that the user writes more in the callback, and you want to keep the writes ordered

txdv replied:
Ha, saghul suggested that I should do it after the connect callback.

I don't understand this suggestion though.

We issued the connection request first and now you want to execute the writes before the connection request?

saghul replied:
I'd say writes always come _after_connect. Actually this patch makes it impossible to do a write before connect, so it makes things easier to reason about.

trevnorris replied:
@txdv 

> We issued the connection request first and now you want to execute the writes before the connection request?

What I understood from @piscisaureus was that he believed the writes should occur just before the `uv_connect_cb` runs (hopefully he wouldn't mind jumping in and giving some feedback ;-). 

@saghul 

> I'd say writes always come _after_connect.

Do you mean _after_ the `uv_connect_cb` has run, or just after the connection has been made?

saghul replied:
On 09/25/2014 06:45 PM, Trevor Norris wrote:

> In src/win/pipe.c:
> 
> > @@ -1742,6 +1779,11 @@ void uv_process_pipe_connect_req(uv_loop_t\* loop, uv_pipe_t\* handle,
> >      req->cb(req, uv_translate_sys_error(err));
> >    }
> > -  if (!err)
> > -    uv__write_callbacks((uv_stream_t*) handle);
> > -  else
> > -    uv__stream_flush_write_queue((uv_stream_t*) handle, UV_ECANCELED);
> 
> @txdv https://github.com/txdv
> 
> ```
> We issued the connection request first and now you want to execute
> the writes before the connection request?
> ```
> 
> What I understood from @piscisaureus https://github.com/piscisaureus
> was that he believed the writes should occur just before the
> |uv_connect_cb| runs (hopefully he wouldn't mind jumping in and giving
> some feedback ;-).
> 
> @saghul https://github.com/saghul
> 
> ```
> I'd say writes always come *after*connect.
> ```
> 
> Do you mean /after/ the |uv_connect_cb| has run, or just after the
> connection has been made?

I mean just after the connection callback was called. Unix does it this
way, and it doesn't make much sense to make it the other way around
IMHO, since it's not possible to write anything before being connected,
so getting the status of those operations in the same order is expected.

txdv replied:
request callbacks for a particular handle have to be always called back in the the order the requests were issued.

trevnorris replied:
okie dokie. that's the way I thought it should be done as well. was just passing on info from a conversation Bert and I had. :)

trevnorris commented on src/win/stream.c:
```diff
@@ -126,9 +126,21 @@ int uv_write(uv_write_t* req,
+      !(handle->flags & UV_HANDLE_WRITABLE))
+    return UV_EINVAL;
+
+  if (handle->type == UV_TTY && ((uv_tty_t*) handle)->handle == INVALID_HANDLE_VALUE)
```
style: passes 80 columns.

txdv replied:
i addressed this one

trevnorris commented on src/win/tcp.c:
```diff
@@ -876,6 +881,42 @@ int uv_tcp_write(uv_loop_t* loop,
+  if (handle->flags & UV_HANDLE_WRITABLE) {
+    req->bufs = (uv_buf_t*) bufs;
+    return uv_tcp_write_req(req);
+  } else {
```
Drop the `} else {`. The previous `if` is returning.

trevnorris commented on src/win/tcp.c:
```diff
@@ -1116,8 +1159,14 @@ void uv_process_tcp_connect_req(uv_loop_t* loop, uv_tcp_t* handle,
+  if (!err)
+    uv__write_callbacks((uv_stream_t*) handle);
+  else
+    uv__stream_flush_write_queue((uv_stream_t*) handle, UV_ECANCELED);
```
Refer to https://github.com/joyent/libuv/pull/1487/files#r18003994

txdv committed:
remove } else { }

the if before returns, no need for nastnig

txdv committed:
use uv_is_closing

txdv committed:
flatten code

txdv committed:
style fix

saghul commented on src/win/stream-inl.h:
```diff
@@ -53,4 +53,50 @@ INLINE static void uv_connection_init(uv_stream_t* handle) {
+
+    req = QUEUE_DATA(q, uv_write_t, queue);
+
+    if (req != NULL && req->cb != NULL)
```
why would req be NULL?

txdv replied:
ill just leave an assert, ok?

saghul replied:
It's really not possible for it to be NULL. You are pointing somewhere inside the write request and QUEUE_DATA will just move the pointer a bit backwards to find the start of the structure. No you're guaranteed to point somewhere in memory. QUEUE_DATA is actually equivalent to container_of.

saghul commented on src/win/tcp.c:
```diff
@@ -813,19 +816,21 @@ int uv_tcp_getpeername(const uv_tcp_t* handle,
-                 const uv_buf_t bufs[],
-                 unsigned int nbufs,
-                 uv_write_cb cb) {
+int uv_tcp_write_req(uv_write_t* req) {
```
This function (and it's equivalent for pipes) should not return errors anymore. We should defer them to the callback, specially now that we might actually do the writes later. This makes it consistent with the general model we (try to) follow: parameter checks return immediately, runtime failures go on the callback. Have a look at what `uv__write` does on Unix.

txdv replied:
We actually do it here: https://github.com/txdv/libuv/blob/write_queue/src/win/stream-inl.h#L67-L77

This was the best way for minimum code change to get the same effect you described.

Do you really want me to rewriting everything?

saghul replied:
That's not correct, because if the request succeeds the callback is queued, but it fails with an unexpect error we'd run the callback immediately, thus breaking the order guarantee.

I know this fix opened the can of worms, but since we are here, let's work to make it right :-)

txdv replied:
That freaking order, you are right.

txdv replied:
Well instead of rewriting the entire function, I'll just add the queuing logic [here](https://github.com/txdv/libuv/blob/write_queue/src/win/stream-inl.h#L79)?

saghul replied:
I'd rather not, because the fix is two fold: when you do `uv_write` _after_ being connected, failures should be reported in the callback. If you fix it here we solve two bugs for the price of one :-)

txdv committed:
rename, saghuls suggestions

saghul commented on src/win/stream-inl.h:
```diff
@@ -53,4 +53,50 @@ INLINE static void uv_connection_init(uv_stream_t* handle) {
+      else
+        assert(0);
+
+      if (err && req->cb != NULL)
```
See my comment here: https://github.com/joyent/libuv/pull/1487/files#r18018384. This shouldn't be necessary.

txdv committed:
Use NULL instead of 0

saghul commented on src/win/stream-inl.h:
```diff
@@ -53,4 +53,50 @@ INLINE static void uv_connection_init(uv_stream_t* handle) {
+
+    req = QUEUE_DATA(q, uv_write_t, queue);
+
+    if (req != NULL) {
```
I don't think req can ever be NULL.

txdv committed:
80 chars80 chars80 chars80 chars80 chars80 chars80 chars80 chars

saghul commented on src/win/stream-inl.h:
```diff
@@ -53,4 +53,50 @@ INLINE static void uv_connection_init(uv_stream_t* handle) {
+      if (stream->type == UV_TCP)
+        err = uv_tcp_write_req(req);
+      else if (stream->type == UV_NAMED_PIPE)
+        err = uv_pipe_write_impl_req(req);
```
After looking at this again, I think I'd rather see this function split in 2: one on tcp.c which works for TCP and another one on pipe.c which does it's job for pipes.

Yes, they'll almost be the same, but there is no real code to be reused here, just a loop iteration.

txdv replied:
We could do it with a callback, like I did before?

saghul replied:
I'd rather not. It looks like we are trying to unify stream handling when it's not really possible here. We don't even do this for TTYs.

txdv replied:
Yeah, because there is no connect on TTY, there is no step between the init and the write.

txdv replied:
why dont you want the callback based method?

saghul replied:
> why dont you want the callback based method?

I'm not against such an interface in general terms, but yes in this
case, because it tries to pretend there is a consistent way of writing
to streams, when there isn't.

txdv replied:
It is internal anyway and you just have to dispatch the queued requests with an implementation specific callback...

saghul commented:
Left a second round of comments. As a general comment, please rename all internal functions you are touching to use the internal convention of `uv__foo`.

txdv committed:
no need for req != NULL, just leave an assert

txdv commented:
you said to rename the internal functions, but `uv_pipe_write_impl` was an internal function before, should I rename it as well?

saghul commented:
> you said to rename the internal functions, but uv_pipe_write_impl was an internal function before, should I rename it as well?

Since you're moving related code around, please do.

txdv committed:
rename my newly added functions.

txdv committed:
rename some other internal functions (the once I touched).

trevnorris commented on src/win/stream-inl.h:
```diff
@@ -53,4 +53,52 @@ INLINE static void uv_connection_init(uv_stream_t* handle) {
+      assert(0);
+
+    if (err && req->cb != NULL)
+      (*req->cb)(req, uv_translate_sys_error(err));
```
Sorry, I'm confused why it's `(*req->cb)(...)` and not just `req->cb(...)`. Mind explaining?

txdv replied:
Nope, it is bad, I'm changing it to req->cb.

saghul commented:
We are going to need a big comment in the code explaining what and when we queue data. I see that in some cases (when we don't queue) we just assign the buffers passed by the user to `req->bufs` and call `uv__tcp_write_req` for example, but when we queue the request we allocate memory for it, which we then need to free.

We need to find a way to make this obvious and easy to follow. One possibility would be to modify `uv__tcp_write_req` so that it gets the buffers passed as a parameter. In one case they would come stratight from the user, in another they would come from the request itself, and freed right afterwards.

txdv commented:
Freeing is not the difficult part, its a simple check against handle->buf == handle->smbuf.

The problem I'm facing right now is guaranteeing that the callbacks are called in order.

It is tedious for examples like these:

```
write_req 1 // succeeds (but is queued, not immediately executed)
write_req 2 // fails immediately for some reason (the os returns enomem or whatever) (we have already something in the queue, so we can't just call the callback, we need to guarantee the callback order, but we need to increment the pending requests counter and such).
write_req 3 // succes (is queued)
write req 1 finishes, write_req 2 failure code has to be called
```

saghul commented:
> Freeing is not the difficult part, its a simple check against handle->buf == handle->smbuf.

Nope. Imagine I want to write 5 buffers. If I do it while connecting we'll malloc the space and memcpy the structures to req->bufs. If I do the same once the connection has been made, however, you just assign req->bufs to whatever the used passes to us. This makes the code complicated and potentially error prone and needs to be addressed.

saghul commented:
> The problem I'm facing right now is guaranteeing that the callbacks are called in order.

Isn't this enough? (example for uv__tcp_write_req)

```
SET_REQ_ERROR(req, WSAGetLastError());
uv_insert_pending_req(loop, (uv_req_t*)req);
```

So, always queue the req with its result.

txdv commented:
> Nope. Imagine I want to write 5 buffers. If I do it while connecting we'll malloc the space and memcpy the structures to req->bufs. If I do the same once the connection has been made, however, you just assign req->bufs to whatever the used passes to us. This makes the code complicated and potentially error prone and needs to be addressed.

In uv__run_write_callbacks we kinda pass the queued writes (before connect was issued) to the IOCP and we can free the bufs afterwards

saghul commented:
> In uv__run_write_callbacks we kinda pass the queued writes (before connect was issued) to the IOCP and we can free the bufs afterwards

I know, but we shouldn't use req->bufs to store some pointers which need to be freed and some which don't. (assuming nbufs > 4)

saghul commented:
@txdv ping

txdv commented:
I'm back from my holiday on wednesday. I won't have access to a windows machine until then.

saghul commented:
Kewl, thanks for the heads up!
On Oct 6, 2014 7:56 PM, "Andrius Bentkus" notifications@github.com wrote:

> I'm back from my holiday on wednesday. I won't have access to a windows
> machine until then.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/joyent/libuv/pull/1487#issuecomment-58059454.

txdv commented:
I don't get it, do you want me to create a separate field for the bufs which libuv allocates?

It is very simple to keep track of them. We are no connected, libuv mallocs them.

When we connect we pass the queued requests to IOCP and free the buffers.

Why can't we have them in the same field?

txdv committed:
brainfart

txdv committed:
queue pending requesst like saghul suggested

txdv commented:
Review please again, I addressed all comments but your last @saghul.

saghul commented on src/win/stream-inl.h:
```diff
@@ -53,4 +53,54 @@ INLINE static void uv_connection_init(uv_stream_t* handle) {
+
+    req = QUEUE_DATA(q, uv_write_t, queue);
+
+    assert(req != NULL);
```
no need for this assert

saghul commented on src/win/stream-inl.h:
```diff
@@ -53,4 +53,54 @@ INLINE static void uv_connection_init(uv_stream_t* handle) {
+    q = QUEUE_HEAD(&stream->write_queue);
+    QUEUE_REMOVE(q);
+
+    req = QUEUE_DATA(q, uv_write_t, queue);
```
no need for this assert

txdv replied:
this one is addressed, you commented on the wrong line so this doesn't get removed

saghul commented on src/win/stream.c:
```diff
@@ -126,19 +126,35 @@ int uv_write(uv_write_t* req,
   int err;
 
-  if (!(handle->flags & UV_HANDLE_WRITABLE)) {
+  if (handle->flags & UV_HANDLE_CLOSED || uv_is_closing((uv_handle_t*) handle))
```
uv_is_closing already checks for UV_HANDLE_CLOSED.

saghul commented on src/win/stream.c:
```diff
@@ -126,19 +126,35 @@ int uv_write(uv_write_t* req,
-  }
+
+  if (handle->type == UV_NAMED_PIPE &&
+      !((uv_pipe_t*) handle)->connect_req &&
```
This looks pretty ugly. Lets add connect_req to uv_stream_t, event if it's not used in ttys.

txdv replied:
To the windows struct or to the global struct?

saghul replied:
The windows one.

saghul commented on src/win/stream.c:
```diff
@@ -126,19 +126,35 @@ int uv_write(uv_write_t* req,
+  if (handle->type == UV_TCP &&
+      !((uv_tcp_t*) handle)->connect_req &&
+      !(handle->flags & UV_HANDLE_WRITABLE))
+    return UV_EINVAL;
```
ENOTCONN?

txdv replied:
if the guard _handle->type == UV_TCP_ fails we have to return EINVAL, while we have to return ENOTCONN on the rest of the statement?

saghul replied:
I'd say so.

txdv replied:
my comment doesn't make sense, ignore it.

txdv replied:
Should I add a test for this? Like do a write without connecting? I don't know if the unix behavior is the same...

saghul commented on src/win/stream.c:
```diff
@@ -126,19 +126,35 @@ int uv_write(uv_write_t* req,
+      err = uv__pipe_write(loop, req, (uv_pipe_t*) handle, bufs, nbufs, cb);
       break;
     case UV_TTY:
+      if (!(handle->flags & UV_HANDLE_WRITABLE))
```
Unrelated / tangentially related change?

txdv replied:
I removed "if (!(handle->flags & UV_HANDLE_WRITABLE)) {" before, so I thought I would add it here...

txdv replied:
Which was a mistake, I grouped it together with the others again.

saghul commented on src/win/tcp.c:
```diff
@@ -760,6 +762,7 @@ static int uv_tcp_try_connect(uv_connect_t* req,
   } else if (UV_SUCCEEDED_WITH_IOCP(success)) {
     /* The req will be processed with IOCP. */
     handle->reqs_pending++;
+    handle->connect_req = req;
```
connect_req also needs to be set if we succeeded without IOCP (above)

txdv replied:
It is not required for this patch, but I'll add it because this way it is more consistent, and some other code might rely on connect_req having the right value.

saghul replied:
If we don't end up using handle-<connect_req in this patch, then we shouldn't add it.

txdv replied:
Wait wait, we need it if we connect with IOCP.

Look at this:

```
+  if (handle->type == UV_NAMED_PIPE &&
+      !handle->connect_req &&
+      !(handle->flags & UV_HANDLE_WRITABLE))
+    return UV_ENOTCONN;
```

We set connect_req when we are connecting with IOCP, if we connect immediately, then the writable flag is set immediatel so we don't need to connect_req for this guard to kick in. That is what I am saying, we don't really need to touch connect_req if we succeed without iocp.

saghul replied:
> We set connect_req when we are connecting with IOCP, if we connect immediately, then the writable flag is set immediatel so we don't need to connect_req for this guard to kick in. That is what I am saying, we don't really need to touch connect_req if we succeed without iocp.

The writable flag is not set immediately. It's set after `uv_process_tcp_connect_req`is called, at the begining of each loop iteration.

txdv replied:
Yeah I figured that myself now too. It looks like it is right now, if we are not connecting and it is not writable, return enotconn.

saghul commented on src/win/tcp.c:
```diff
@@ -813,19 +816,21 @@ int uv_tcp_getpeername(const uv_tcp_t* handle,
+  handle = (uv_tcp_t*) req->handle;
+  bufs = req->bufs;
+  nbufs = req->nbufs;
+  cb = req->cb;
```
Unused.

txdv replied:
For some reason when compiling with make it still doesn't complain about it. I thought we had a flag added or something.

saghul commented on src/win/tcp.c:
```diff
@@ -876,6 +881,42 @@ int uv_tcp_write(uv_loop_t* loop,
+  if (nbufs > ARRAY_SIZE(req->bufsml)) {
+    req->bufs = malloc(size);
+    if (req->bufs == NULL)
+      return UV_ENOMEM;
```
The caller of this function expects a Windows error, not a libuv error.

saghul commented on include/uv-win.h:
```diff
@@ -438,7 +443,8 @@ RB_HEAD(uv_timer_tree_s, uv_timer_s);
     struct { uv_tcp_connection_fields };                                      \
-  };
+  };                                                                          \
+  uv_connect_t* connect_req;
```
add this to UV_STREAM_PRIVATE_FIELDS instead

saghul commented on src/win/stream.c:
```diff
@@ -126,19 +126,35 @@ int uv_write(uv_write_t* req,
+
+  if (handle->type == UV_TTY &&
+      ((uv_tty_t*) handle)->handle == INVALID_HANDLE_VALUE)
+    return UV_EINVAL;
```
Check for writability here instead of below, for consistency.

saghul commented on src/win/stream.c:
```diff
@@ -126,19 +126,35 @@ int uv_write(uv_write_t* req,
+      ((uv_tty_t*) handle)->handle == INVALID_HANDLE_VALUE)
+    return UV_EINVAL;
 
   err = ERROR_INVALID_PARAMETER;
```
Set only on default case.

txdv replied:
Should I remove the assert(0) from the default case?

saghul replied:
No.

txdv replied:
Why would we set it all then?

I don't get it.

saghul replied:
Because asserts are no-ops in release builds.

txdv replied:
Ok, but all the surrounding code does it before, not in the default case. Do you maybe want me to leave it for consistency?

saghul commented on src/win/pipe.c:
```diff
@@ -1428,14 +1411,66 @@ static int uv_pipe_write_impl(uv_loop_t* loop,
+
+  memset(&req->overlapped, 0, sizeof(req->overlapped));
+
+  if (handle->handle != INVALID_HANDLE_VALUE) {
```
Hum. Not having a handle is basically like not having a fd, that returns EBADF on Unix. Now, weren't we supposed to check if connect_req is not NULL and queue in that case?

Also, this needs to check if there were previously queued writes, because in that case we need to append to the queue:  create the pipe, write some reqs (queued), uv_pipe_connect, write some more reqs (need to be queued, not immediately written or queue order is fucked up)

saghul commented on src/win/stream-inl.h:
```diff
@@ -53,4 +53,54 @@ INLINE static void uv_connection_init(uv_stream_t* handle) {
 }
 
 
+INLINE static void uv__run_write_callbacks(uv_stream_t* stream) {
```
I still don't like what we do here. Please have some uv__tcp_do_write and uv__pipe_do_write instead.

saghul commented:
> I don't get it, do you want me to create a separate field for the bufs which libuv allocates?
> 
> It is very simple to keep track of them. We are no connected, libuv mallocs them.
> 
> When we connect we pass the queued requests to IOCP and free the buffers.
> 
> Why can't we have them in the same field?

The problem I have is that req->bufs can point to 3 types of things: 
- user allocated memory (when we write immediately)
- libuv malloc-ed memory (when nbufs > 4)
- req->bufsml(when nbufs <= 4 and write is queued)

Distinguishing between the last 2 is easy, but I worry the first case will give us trouble.

What I suggested is to pass bufs and nbufs to the function that does the immediate write, this way the parameters will always be untouched, but can come from 2 places:
- user passed structures
- structures which we hold in the req

saghul commented on src/win/tcp.c:
```diff
@@ -876,6 +881,42 @@ int uv_tcp_write(uv_loop_t* loop,
+  req->nbufs = nbufs;
+  req->cb = cb;
+
+  if (handle->flags & UV_HANDLE_WRITABLE) {
```
This is not enough: example case:
- uv_tcp_connect
- uv_write (writes are queued)
- connect_cb
- uv_write

Here the writable flag is already set and the writes done inside the connect callback would be done before the ones we queued earlier.

txdv replied:
We actually cover that in uv_write@src/win/stream.c:

``` c
   if (handle->type == UV_TCP &&
       !handle->connect_req &&
       !(handle->flags & UV_HANDLE_WRITABLE))
     return UV_ENOTCONN;
```

saghul commented:
Finished another review round.

txdv committed:
No need for this assert.

txdv committed:
uv_is_closing already checks for UV_HANDLE_CLOSED.

txdv committed:
loop doesn't exist, we need to address it via the handle

txdv committed:
Move req_connect field to main stream struct.

Makes the resulting code more readable, because we don't need conversations
any more.

txdv committed:
Line too long.

txdv committed:
return ENOTCONN if we write before connect.

txdv committed:
set connect_req even if we succeed immediately.

txdv committed:
remove unused variable.

txdv committed:
The caller of this function expects a Windows error, not a libuv error.

txdv committed:
consistency

txdv committed:
Set only on default case.

txdv committed:
uv__tcp_do_write and uv__pipe_do_write

I don't like this.

txdv committed:
remove unused variable

txdv committed:
Make the request struct never hold the user specified bufs/nbufs on windows.

txdv commented:
Well, from now on it will never hold user allocated memory, bufs seems to be a private field, so I guess that is ok.

txdv committed:
Keep that assert.

txdv commented on src/win/stream-inl.h:
```diff
@@ -53,4 +53,68 @@ INLINE static void uv_connection_init(uv_stream_t* handle) {
+
+    if (err && req->cb != NULL) {
+      SET_REQ_ERROR(req, WSAGetLastError());
+      uv_insert_pending_req(stream->loop, (uv_req_t*)req);
```
I guess the WSAGetLastError() is wrong too, right? I should just pass the err.

But the main problem with uv_insert_pending_req here is that there can be a race condition.

We have 2 different queues, the pending_req queues (which is used for requests that are already FINISHED by the IOCP and have to be processed(call callbacks, etc.)) and the active_reqs which keeps track of what request are currently in the IOCP queue.

So now when we do a

```
connect(delayed, success)
write(success)
write(fail)
```

libuv now queues the writes even though we are not connected, because IOCP doesn't handle that.

Now the problem is that once we are flushing the queued write requests when we connected successfully and the second write fails, it will get executed before the first write, because we don't do an [insert_pending on IOCP handled writes](https://github.com/txdv/libuv/blob/write_queue/src/win/tcp.c#L860-L864), only [after the GetQeueudCompletionStatus](https://github.com/txdv/libuv/blob/write_queue/src/win/core.c#L351).

This means we have a race condition, because after the connect we immediately insert the write failed in the pending queue while we only do that for the successful write when it succeeds later on.

So write(fail) will be called first.

saghul replied:
Yep, we have a problem there. Maybe we can just set the error in the request and fire the IOCP event, thus making use of the builtin request queuing mechanism. See https://github.com/txdv/libuv/blob/write_queue/src/win/tcp.c#L340

txdv replied:
Are you sure that if we queue something with PostQueuedCompletionStatus before the other write finishes that it won't come out first?

It looks like windows socket implementation calls PostQueuedCompletionStatus itself after the data has been sent, so we if we post it before windows does ... then the order will be bad as well.

saghul replied:
On 10/19/2014 08:44 PM, Andrius Bentkus wrote:

> In src/win/stream-inl.h:
> 
> > +INLINE static void uv__tcp_do_write(uv_stream_t\* stream) {
> > -  uv_write_t\* req;
> > -  int err;
> > -  QUEUE *q;
> >   +
> > -  while (!QUEUE_EMPTY(&stream->write_queue)) {
> > -    q = QUEUE_HEAD(&stream->write_queue);
> > -    QUEUE_REMOVE(q);
> >   +
> > -    req = QUEUE_DATA(q, uv_write_t, queue);
> >   +
> > -    err = uv__tcp_write_req(req, req->bufs, req->nbufs);
> >   +
> > -    if (err && req->cb != NULL) {
> > -      SET_REQ_ERROR(req, WSAGetLastError());
> > -      uv_insert_pending_req(stream->loop, (uv_req_t*)req);
> 
> Are you sure that if we queue something with PostQueuedCompletionStatus
> before the other write finishes that it won't come out before?

No, that probably won't work either... we might need some
write_completed_queue like we have on Unix.

## 

Saúl Ibarra Corretgé
bettercallsaghul.com

txdv replied:
I'll push a commit in the morning with my idea how we can do it.

saghul commented on src/win/stream-inl.h:
```diff
@@ -53,4 +53,68 @@ INLINE static void uv_connection_init(uv_stream_t* handle) {
 }
 
 
+INLINE static void uv__tcp_do_write(uv_stream_t* stream) {
```
Move this one to tcp.c

saghul commented on src/win/pipe.c:
```diff
@@ -1742,6 +1778,11 @@ void uv_process_pipe_connect_req(uv_loop_t* loop, uv_pipe_t* handle,
   }
 
+  if (!err)
+    uv__tcp_do_write((uv_stream_t*) handle);
```
s/tcp/pipe/

saghul commented on src/win/stream-inl.h:
```diff
@@ -53,4 +53,68 @@ INLINE static void uv_connection_init(uv_stream_t* handle) {
+}
+
+
+INLINE static void uv__pipe_do_write(uv_stream_t* stream) {
```
Move this one to pipe.c

txdv committed:
rename to pipe

txdv committed:
move stuff

txdv committed:
set error from functino.

txdv commented on test/test-write-after-connect.c:
```diff
@@ -0,0 +1,153 @@
+
+
+static void tcp_server(uv_connection_cb callback) {
+  ASSERT(0 == uv_tcp_init(&loop, &client.tcp));
```
this line is not needed

txdv commented on src/win/internal.h:
```diff
@@ -134,7 +134,7 @@ int uv_tcp_listen(uv_tcp_t* handle, int backlog, uv_connection_cb cb);
 int uv_tcp_read_start(uv_tcp_t* handle, uv_alloc_cb alloc_cb,
     uv_read_cb read_cb);
-int uv_tcp_write(uv_loop_t* loop, uv_write_t* req, uv_tcp_t* handle,
+int uv__tcp_write(uv_loop_t* loop, uv_write_t* req, uv_tcp_t* handle,
```
should i put this somewhere else? uv__ at the end or something?

saghul replied:
You can leave it there.

txdv commented on src/win/tcp.c:
```diff
@@ -1116,8 +1179,14 @@ void uv_process_tcp_connect_req(uv_loop_t* loop, uv_tcp_t* handle,
   } else {
     err = GET_REQ_SOCK_ERROR(req);
   }
+
```
new line? should I add it?

saghul replied:
On 10/20/2014 12:01 AM, Andrius Bentkus wrote:

> In src/win/tcp.c:
> 
> > @@ -1116,8 +1179,14 @@ void uv_process_tcp_connect_req(uv_loop_t\* loop, uv_tcp_t\* handle,
> >    } else {
> >      err = GET_REQ_SOCK_ERROR(req);
> >    }
> > +
> 
> new line? do should I add it?

New line is ok if it helps readability.

txdv commented on src/win/tcp.c:
```diff
@@ -1116,8 +1179,14 @@ void uv_process_tcp_connect_req(uv_loop_t* loop, uv_tcp_t* handle,
     err = GET_REQ_SOCK_ERROR(req);
   }
+
   req->cb(req, uv_translate_sys_error(err));
```
shouldn't we check if req->cb != NULL, becaue something like uv_connect(addr, NULL) seems now feasible?

saghul replied:
On 10/20/2014 12:02 AM, Andrius Bentkus wrote:

> In src/win/tcp.c:
> 
> > @@ -1116,8 +1179,14 @@ void uv_process_tcp_connect_req(uv_loop_t\* loop, uv_tcp_t\* handle,
> >    } else {
> >      err = GET_REQ_SOCK_ERROR(req);
> >    }
> > +
> >    req->cb(req, uv_translate_sys_error(err));
> 
> shouldn't we check if req->cb != NULL, becaue something like
> uv_connect(addr, NULL) seems now feasible.

It makes no sense to not have a callback in uv_tcp|pipe_connect. We
should give EINVAL in case NULL was passed.

txdv replied:
My vision is to have code like this:

```
uv_connect(..., NULL)
uv_write(..., NULL)
uv_shutdown(..., CALLBACK);
```

It will queue the connect, the write and the shutdown immediately and call CALLBACK with an error if the previous steps failed. It kinda make sense if you consider that we can immediately queue the writes without really connecting!

Also it make libuv usage far more easier for C users.

saghul replied:
Unix seems to allow it, so lets go with that.

txdv replied:
This would allow us to connect, write and shutdown without a need for a callback!

Everything would be queued and we wouldn't need callbacks.

Though making shutdown work correctly in this scenario will need an additional patch. I don't think it belongs in here, because this one is already so fat. I think this goes along the spirit of uv_shutdown which is queued and executed somewhere down the line.

txdv commented on src/win/pipe.c:
```diff
@@ -1180,40 +1184,19 @@ static void uv_queue_non_overlapped_write(uv_pipe_t* handle) {
+  uv_loop_t* loop = req->handle->loop;
+  uv_pipe_t* handle = (uv_pipe_t*) req->handle;
+  uv_stream_t* send_handle = req->send_handle;
+  uv_write_cb cb = req->cb;
```
should i immediately initialize the variables, or just do a variable decleration and an initialization later on. This looks more compact.

saghul replied:
On 10/20/2014 12:08 AM, Andrius Bentkus wrote:

> In src/win/pipe.c:
> 
> > ## \-  }
> > 
> > ## \-  assert(handle->handle != INVALID_HANDLE_VALUE);
> > -  uv_req_init(loop, (uv_req_t*) req);
> > -  req->type = UV_WRITE;
> > -  req->handle = (uv_stream_t*) handle;
> > -  req->cb = cb;
> > -  req->ipc_header = 0;
> > -  req->event_handle = NULL;
> > -  req->wait_handle = INVALID_HANDLE_VALUE;
> > -  memset(&req->overlapped, 0, sizeof(req->overlapped));
> > -  uv_loop_t\* loop = req->handle->loop;
> > -  uv_pipe_t\* handle = (uv_pipe_t*) req->handle;
> > -  uv_stream_t\* send_handle = req->send_handle;
> > -  uv_write_cb cb = req->cb;
> 
> should i immediately initialize the variables, or just do a variable
> decleration and an initialization later on.

Generally we do it in 2 steps.

txdv commented on src/win/tcp.c:
```diff
@@ -813,19 +817,17 @@ int uv_tcp_getpeername(const uv_tcp_t* handle,
+  uv_tcp_t* handle;
+
+  loop = req->handle->loop;
+  handle = (uv_tcp_t*) req->handle;
```
Here i am declaring it first and then initializing it. Maybe I should merge them like in the pipe exmaple?

txdv replied:
two step it is i guess

txdv committed:
remove not needed client initialization

txdv commented on src/win/pipe.c:
```diff
@@ -1724,6 +1758,30 @@ void uv_process_pipe_accept_req(uv_loop_t* loop, uv_pipe_t* handle,
+
+    err = uv__pipe_write_impl_req(req, req->bufs, req->nbufs);
+
+    if (err && req->cb != NULL) {
```
Should I use _req->cb_ or _req->cb != NULL_?

I think the former is the correct one, because that is most commonly used throughout the code base.

Though it is used inconsistently. We kinda have to make a call on this for future flaming @saghul 

I'm kinda for if (req->cb)

saghul replied:
On 10/20/2014 10:50 AM, Andrius Bentkus wrote:

> In src/win/pipe.c:
> 
> > @@ -1724,6 +1758,30 @@ void uv_process_pipe_accept_req(uv_loop_t\* loop, uv_pipe_t\* handle,
> >  }
> > 
> > +INLINE static void uv__pipe_do_write(uv_stream_t\* stream) {
> > -  uv_write_t\* req;
> > -  int err;
> > -  QUEUE *q;
> >   +
> > -  while (!QUEUE_EMPTY(&stream->write_queue)) {
> > -    q = QUEUE_HEAD(&stream->write_queue);
> > -    QUEUE_REMOVE(q);
> >   +
> > -    req = QUEUE_DATA(q, uv_write_t, queue);
> >   +
> > -    err = uv__pipe_write_impl_req(req, req->bufs, req->nbufs);
> >   +
> > -    if (err && req->cb != NULL) {
> 
> Should I use /req->cb/ or /req->cb != NULL/?
> 
> I think the former is the correct one, because that is most commonly
> used throughout the code base.

Both are correct, so go with the one used more consistently throughout
the code.

txdv replied:
I'll go with req->cb because it is less verbose.

bnoordhuis usually goes with req->cb != NULL, the windows guy went with req->cb, i go with req->cb because it is more often used and I use it in other code base as well

txdv committed:
two step initialization

txdv committed:
Add callback guard

txdv committed:
codestyle consistency

txdv committed:
initialzie pipe handle even the though the callback is not set

this will allow us to handle scenarios like

uv_connect(..., NULL)
uv_write(..., NULL)
uv_shutdown(..., CALLBACK)

which will just queue the write and shutdown and execute them
once everything is done.

txdv committed:
add missing semi colon

txdv commented:
As I was writing the patch for the correct ordering, I saw that I forgot to count the pending write requests.

txdv committed:
codestyle

txdv committed:
Count correctly.

Less or more, the do_write implementation is still faulty

txdv committed:
ensure order for tcp write requests

txdv commented:
This is my approach. Works currently only for tcp.

The code duplicated is now huge because we have tcp_do_write and pipe_do_write, can I write a common abstraction with both now?

txdv committed:
do the same for pipes

txdv committed:
use out of memory in pipe as well

txdv committed:
add forgotten size

saghul commented on include/uv-win.h:
```diff
@@ -414,13 +419,16 @@ RB_HEAD(uv_timer_tree_s, uv_timer_s);
-  };
+  };                                                                          \
+  uv_connect_t* connect_req;                                                  \
+  uv_write_t* prev_write_req;
```
You don't seem to use this one.

txdv replied:
Because I assigned the last request badly in the write function.

saghul commented on src/win/stream-inl.h:
```diff
@@ -53,4 +53,94 @@ INLINE static void uv_connection_init(uv_stream_t* handle) {
+      SET_REQ_ERROR(req, err);
+      /* if we had no succesful write before, we can can just queue the
+         request immediately, otherwise we need it to keep in the write queue
+         and put it the pending queue when the successful write gets handled */
```
what would happen if all write requests fail?

txdv replied:
Until there is a successful write, it just puts it immediately in the pending queue. So if all requests fail, it puts them all immediately in the pending queue which then gets handled.

saghul commented:
The code looks overly complicated right now, specially with the manual counter management. Also, prev_write_req seems to be only assigned in a single place, with a pointer pointing to the contained request, so looks like it could be removed and compare requests instead (I might be missing something here).

I need to read this more thoroughly when I find the time.

txdv committed:
remember last write request correctly

txdv commented:
When will you find the time? Any approximation?

saghul commented:
As soon as I can. I'll try to do so in the upcoming days, since I'm leaving for a conference + vacation on friday. Building and destroying the mental model for this is quite time consuming :-(

Also, FWIW, I'm planning on dropping this from the 1.0.0 release. It's pretty late in the process, and I don't want us to introduce such a big patch as of now. I think neither of us knew it was going to grow this big / involved, and I'd rather not rush things. Once we are happy with it it shall land in master.

txdv commented:
I don't really care about it landing in 1.0.0.

I actually want you to release 1.0.0 ASAP so I can start doing breaking changes.

txdv committed:
bad variable name

txdv committed:
boil out common code

txdv committed:
brainfart

txdv commented:
> As soon as I can.

Yeah I figured, I hoped for a quantified answer.

txdv committed:
use ssize_t instead of int

txdv committed:
Use unsigned version of size_t

txdv commented:
@saghul you still alive?

saghul commented:
Still out, at about 600ms from any continent.

On Tue, Nov 4, 2014 at 4:50 AM, Andrius Bentkus notifications@github.com
wrote:

> @saghul https://github.com/saghul you still alive?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/joyent/libuv/pull/1487#issuecomment-61615891.

## 

/Saúl
bettercallsaghul.com

txdv commented:
Best time to ask for a 1on1 quake match I guess. I'll just start writing the patch for queuing the shutdown request on top of this.

When are you comming back?

txdv commented:
no resolution

saghul commented:
I still planned to continue reviewing this after 1.0, I just got back on Sunday.

saghul commented on src/win/pipe.c:
```diff
@@ -1428,14 +1416,49 @@ static int uv_pipe_write_impl(uv_loop_t* loop,
+  if (handle->handle != INVALID_HANDLE_VALUE)
+    return uv__pipe_write_impl_req(req, bufs, nbufs);
+  else
+    return vu__queue_write_queue((uv_stream_t*) handle, req, bufs, nbufs);
```
typo: uv_queue

saghul commented on src/win/stream-inl.h:
```diff
@@ -52,5 +52,120 @@ INLINE static void uv_connection_init(uv_stream_t* handle) {
   handle->shutdown_req = NULL;
 }
 
+INLINE static int vu__queue_write_queue(uv_stream_t* handle,
```
uv

saghul replied:
I don't think this needs to be inline, the compiler should know.

saghul commented:
Keeping the linked list with requests seems to work (I haven't run the code yet though), but I'd like to explore another approach to the solution, which is actually analogous to what we do on Unix: write one request at a time. The idea qould be to only do the actual write if there are no other queued writes, and when one write completed do the write for the next.

I talked with @piscisaureus and he told me that letting the kernel handle all the queuing might not always be a good idea, because there some uknown threshold and after reaching it the kernel won't report sucess untill all write operations have been completed.

The open question is, of course, if there is a performance hit.

txdv commented:
> after reaching it the kernel won't report sucess untill all write operations have been completed.

Why **all**?

saghul commented:
> Why all?

I wish I knew... This is probably what was happening in a test case where we queued 100k write requests and the whole thing would even blow up from time to time.

txdv commented:
100K writes? It is kinda a big threshold, I would be more concerned with returning NOMEM in the right places.... But I guess we could try to benchmark it. Are you saying that this won't be merged until I do the benchmark with queuing and only processing N processes at once? Where N might or might not be 1 as well?

saghul commented:
> 100K writes? It is kinda a big threshold, I would be more concerned with returning NOMEM in the right places.... But I guess we could try to benchmark it. 

I don't know where the threshold really is, I do remember seeing weird things with that load, that's all.

> Are you saying that this won't be merged until I do the benchmark with queuing and only processing N processes at once? Where N might or might not be 1 as well?

Yes, kind of. The current approach (assuming it covers all cases) is pretty hard to follow / maintain. I'd like to see how the other approach performs before taking any decission here. If there is no performance penalty I'd favor that one. (N being 1)

Also, @piscisaureus, I summon you!

piscisaureus commented:
The issue on windows is that if you queue a lot of data to the kernel, completion of async writes start to behave differently. Effectively instead of getting a completion notification for a write as soon its data has been sent (which is what normally happens), all completion notifications seem to be held back until the kernel send buffer is completely empty.

This effect seems to be triggered by queuing a lot of send data to the kernel (in big chunks), not necessarily a lot of write operations per se.

I recall there was a node issue where the problem was demonstrated in detail but I couldn't find it, although https://github.com/joyent/node/pull/1256 seems to reference it also.

When I first saw it with my own eyes I was shocked and surprised. I don't know why this happens.

The trick is going to be to limit the number of simultaneously queued requests / bytes (per socket) that is dispatched to the kernel; if the user writes more it'll be buffered in user space until a kernel-queued write returns.

So let me draft a reasonable proposal.
- If there are less than 2 write requests queued to the kernel, queue a new write to the kernel also.
- If there is less than 128k of write data queued to the kernel queue a new write to the kernel also.
- Otherwise queue in userland.

txdv commented:
You could have told me that right in the beginning of the patch... I could have worked immediately on this.
